server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -1359,6 +1360,11 @@ public GatedCloseable<IndexCommit> acquireLastIndexCommit(boolean flushFirst) th
         }
     }
 
+    public void finalizeReplication(SegmentInfos infos, long seqNo) throws IOException {
```

Comment:
It seems like we are adding the method to be called by everyone but only supported for `NRTReplicationEngine`. May we could open up the `getEngine()` and `SegmentReplicationTarget` could make the decision if `updateSegments` should be called?

Reply:
For now I'm still concerned about opening up access of engine beyond IndexShard

Reply:
I think we can still do better than that:

```
public interface ReplicationEngine {
    void updateSegments(final SegmentInfos infos, long seqNo) throws IOException;
}
```

Than IndexShard can do:
```
public Optional<ReplicationEngine> getReplicationEngine () {
    if (getEngine() instanceof ReplicationEngine)) {
        return Optional.of((ReplicationEngine)getEngine());
    } else {
       return Optional.empty();
    }
}
```

The main idea here: if something is not implemented, it should not be called. In this case, if engine does not support replication - the `getReplicationEngine()` returns empty and the caller won't be able to make a call.

Reply:
Wondering if we really need a new Engine interface, does the current Engine interface not perform anything that has to do with replication already?

Reply:
@reta I understand what the interface intends to do, what I have a concern on is the interface looks incomplete to me. I am not quite sure if `void updateSegments(final SegmentInfos infos, long seqNo) throws IOException;` is the only thing a `ReplicationEngine` should do

Reply:
@Bukhtawar 👍

Reply:
`ReplicationEngine` is a regular Engine implementation and right now the only additional functionality of updating & refreshing on an externally provided SegmentInfos in `NRTReplicationEngine`.  @Bukhtawar what additional functionality do you think is missing?  We could add a default implementation in Engine that throws a `UnsupportedOperationException` and only implement it in `NRTReplicationEngine` but I like the isolated interface.

Reply:
I would revisit the `Engine` interface and see what can be extracted into a separate `ReplicationEngine ` interface and  decide on the split. Based on what you mention @mch2 we can also create a functional interface and call it a `SegmentUpdateService` instead so that it appears less overloaded and performs the limited responsibility it specialises in/is tasked with.

```
public interface SegmentUpdateService {
    
    @FunctionalInterface
    void updateSegments(final SegmentInfos infos, long seqNo) throws IOException;
}
```

Reply:
I like this.  I think we can also remove the `finalizeReplication` method here and invoke `indexShard.getSegmentUpdateService().updateSegments` directly from the target class.

```java
    public SegmentUpdateService getSegmentUpdateService() {
        final Engine engine = getEngine();
        assert engine instanceof SegmentUpdateService;
        return (SegmentUpdateService) engine;
```

Reply:
i agree that the current `ReplicationEngine` interface seems out of place and incomplete. Why not use `NRTReplicationEngine` directly?

```
public Optional<NRTReplicationEngine> getReplicationEngine () {
    if (getEngine() instanceof NRTReplicationEngine)) {
        return Optional.of((NRTReplicationEngine)getEngine());
    } else {
       return Optional.empty();
    }
}
```